lib/util/traverser.js

```diff
@@ -1,34 +1,143 @@
-    "trailingComments"
-]);
+/**
+ * Do nothing.
+ * @returns {void}
+ */
+function noop() {
+
+    // do nothing.
+}
+
+/**
+ * Check whether the given value is an ASTNode or not.
+ * @param {any} x The value to check.
+ * @returns {boolean} `true` if the value is an ASTNode.
+ */
+function isNode(x) {
+    return x !== null && typeof x === "object" && typeof x.type === "string";
+}
 
 /**
- * Wrapper around an estraverse controller that ensures the correct keys
- * are visited.
- * @constructor
+ * The traverser class to traverse AST trees.
  */
-class Traverser extends estraverse.Controller {
-    traverse(node, visitor) {
-        visitor.fallback = Traverser.getKeys;
-        return super.traverse(node, visitor);
+class Traverser {
+    constructor() {
+        this._current = null;
+        this._parents = [];
+        this._skipped = false;
+        this._broken = false;
+        this._visitorKeys = null;
+        this._enter = null;
+        this._leave = null;
+    }
+
+    /**
+     * @returns {ASTNode} The current node.
+     */
+    current() {
+        return this._current;
+    }
+
+    /**
+     * @returns {ASTNode[]} The ancestor nodes.
+     */
+    parents() {
+        return this._parents.slice(0);
+    }
+
+    /**
+     * Break the current traversal.
+     * @returns {void}
+     */
+    break() {
+        this._broken = true;
+    }
+
+    /**
+     * Skip child nodes for the current traversal.
+     * @returns {void}
+     */
+    skip() {
+        this._skipped = true;
+    }
+
+    /**
+     * Traverse the given AST tree.
+     * @param {ASTNode} node The root node to traverse.
+     * @param {Object} options The option object.
+     * @param {Object} [options.visitorKeys=DEFAULT_VISITOR_KEYS] The keys of each node types to traverse child nodes. Default is `./default-visitor-keys.json`.
+     * @param {Function} [options.enter=noop] The callback function which is called on entering each node.
+     * @param {Function} [options.leave=noop] The callback function which is called on leaving each node.
+     * @returns {void}
+     */
+    traverse(node, options) {
+        this._current = null;
+        this._parents = [];
+        this._skipped = false;
+        this._broken = false;
+        this._visitorKeys = options.visitorKeys || vk.KEYS;
+        this._enter = options.enter || noop;
+        this._leave = options.leave || noop;
+        this._traverse(node, null);
+    }
+
+    /**
+     * Traverse the given AST tree recursively.
+     * @param {ASTNode} node The current node.
+     * @param {ASTNode|null} parent The parent node.
+     * @returns {void}
+     * @private
+     */
+    _traverse(node, parent) {
+        if (!isNode(node)) {
+            return;
+        }
+
+        this._current = node;
+        this._skipped = false;
+        this._enter(node, parent);
+
+        if (!this._skipped && !this._broken) {
+            const keys = this._visitorKeys[node.type] || Traverser.getKeys(node);
```

Comment:
Do we want to have some kind of warning/error if keys don't exist for a node type? I can also see the argument for _not_ doing that (so that it will still parse even when it encounters an unknown node type), but failing to visit a node type silently could be undesirable.

Reply:
This seems like it would cause a lot of noise for existing setups where people are using custom parsers without custom keys.

Reply:
Could we add an ESLint CLI/CLIEngine option to allow users to opt-in to warnings/errors in this case?

Reply:
That feels like a non-trivial amount of work for potentially very little benefit. I more just wanted to make sure we at least think about this and are aware in case something crops up. We can always revisit/add later too, if we feel like we're missing it.

Reply:
I think that we can add `debug("detect unknown node: %s", node.type)` message if people use with `--debug` option.